csrc/fusion.cpp

```diff
@@ -205,8 +205,20 @@ void Fusion::removeVal(Val* val) {
     removeExpr(orig);
   }
 
-  for (Expr* use : unordered_uses(val)) {
-    removeExpr(use);
+  const std::unordered_set<Expr*> exprs = exprs_;
+  std::vector<Expr*> exprs_to_remove;
+  for (Expr* e : exprs_) {
```

Comment:
Having to scan the whole expression list to remove just one `Val` isn't ideal. We may need to later discuss broadly about the contracts around dead code. However, this PR seems to be the fastest and practical way to fix the bug. So LGTM!

Reply:
Yeah I agree. I considered adding another member to track this but ultimately decided it wouldn't be worth maintaining since Val removal is not done all that often.